clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1238,6 +1373,149 @@ namespace {
+      ProofResult
+      CompareUpperOffsets(BaseRange &R, ProofFailure &Cause,
+                          EquivExprSets *EquivExprs,
+                          std::pair<ComparisonSet, ComparisonSet> &Facts,
+                          FreeVariableListTy &FreeVariables) {
+        ProofResult Result =
+            CompareUpperOffsetsImpl(R, Cause, EquivExprs, Facts);
+        if (Result != ProofResult::Maybe)
+          return Result;
+
+        if (CheckFreeVarInExprs(S, R.UpperOffsetVariable, UpperOffsetVariable,
+                                FVPos::DeclaredUpper, FVPos::ObservedUpper,
+                                EquivExprs, FreeVariables)) {
+          Cause = CombineFailures(Cause, ProofFailure::HasFreeVariables);
+          return ProofResult::False;
+        }
+
+        return ProofResult::Maybe;
+      }
+
+      // GetFreeVariables gathers "free variables" in SrcVars.
+      //
+      // Given two variable sets SrcVars and DstVars, and a set of equivalent
+      // sets of Expr EquivExprs. A variable V in SrcVars is *free* if the two
+      // conditions are met:
+      //   1. either (1) V is not an integer type, or (2) there is no set in
+      //   EquivExprs that contains both V and a constant (i.e. an integer-typed
+      //   value); and
+      //   2. for each variable U in DstVars, there is not set in EquivExprs
+      //   that contains both V and U.
+      //
+      // GetFreeVariables returns true if any free variable is found in SrcVars,
+      // and appends the free variables to FreeVariables.
+      static bool GetFreeVariables(Sema &S, const EqualExprTy &SrcVars,
+                            const EqualExprTy &DstVars,
+                            EquivExprSets *EquivExprs,
+                            EqualExprTy &FreeVariables) {
+        bool HasFreeVariables = false;
+
+        // Gather free variables.
+        for (const auto &SrcV : SrcVars) {
+          if (IsEqualToConstant(S, SrcV, EquivExprs))
+            continue;
+          auto It = DstVars.begin();
+          for (; It != DstVars.end(); It++) {
+            if (EqualValue(S.Context, SrcV, *It, EquivExprs))
+              break;
+          }
+
+          // We searched all declared variables and found neither a constant nor
+          // a match for SrcV.
+          if (It == DstVars.end()) {
+            HasFreeVariables = true;
+            FreeVariables.push_back(SrcV);
+          }
+        }
+        return HasFreeVariables;
+      }
+
+      // Check free variables between E1 and E2. Append any found free variables
+      // to FreeVars with each free variable in E1 (resp. E2) having Pos1 (resp.
+      // Pos2).
+      static bool CheckFreeVarInExprs(Sema &S, Expr *E1, Expr *E2, FVPos Pos1,
+                                      FVPos Pos2, EquivExprSets *EquivExprs,
+                                      FreeVariableListTy &FreeVars) {
+        // Do not check free variables in bounds that contain indirect accesses.
```

Comment:
This comment is somewhat confusing since `E1` and `E2` are not necessarily bounds expressions. I suggest changing to "Do not check free variables in expressions that contain indirect accesses.".